File: publish.php
<?php

/**
 * Load podcasts configuration
 */
function load_podcasts_config(string $config_file): array
{
	if (!file_exists($config_file)) {
		error_log("Configuration file not found: $config_file");
		exit(1);
	}

	$config = json_decode(file_get_contents($config_file), true);

	if ($config === null) {
		error_log('Error parsing configuration file: ' . json_last_error_msg());
		exit(1);
	}

	return $config;
}

/**
 * Get tracking file path for a specific podcast
 */
function get_tracking_file(string $podcast_id): string
{
	return "./published_episodes_{$podcast_id}.txt";
}

/**
 * Fetch last episode from podcast feed
 */
function fetch_last_episode(string $feed_url): SimpleXMLElement|false
{
    $feed = simplexml_load_file($feed_url);

    if ($feed === false) {
        error_log('Error fetching feed: ' . print_r(error_get_last(), true));
        exit(1);
    }

    $item = $feed->channel->item[0];

    if ($item === null) {
        error_log('Error fetching last episode: ' . print_r(error_get_last(), true));
        exit(1);
    }

    return $item;
}

/**
 * Publish last episode to social Mastodon
 */
function publish_to_mastodon(SimpleXMLElement $last_episode, string $mastodon_url, string $mastodon_token, string $template): false|string
{
    if (empty($title = $last_episode->title) || empty($link = $last_episode->link)) {
        error_log('Error fetching last episode: ' . print_r(error_get_last(), true));
        exit(1);
    }
	// Extract hashtags from itunes:keywords
	$hashtags = '';
	$itunes_ns = $last_episode->children('http://www.itunes.com/dtds/podcast-1.0.dtd');
	if (isset($itunes_ns->keywords)) {
		$keywords = (string)$itunes_ns->keywords;
		if (!empty($keywords)) {
			$keywords_array = array_map('trim', explode(',', $keywords));
			$hashtags_array = array_map(function($keyword) {
				return '#' . str_replace(' ', '', $keyword);
			}, $keywords_array);
			$hashtags = implode(' ', $hashtags_array);
		}
	}

    $content = str_replace(
	    ['{title}', '{link}', '{hashtags}'],
		[(string)$title, (string)$link, (string)$hashtags],
        $template
    );

    // log content
    echo "Publishing to Mastodon: $content\n";

    $options = array(
        'http' => array(
            'header' => "Content-type: application/x-www-form-urlencoded\r\n" .
                "Authorization: Bearer $mastodon_token\r\n",
            'method' => 'POST',
            'content' => http_build_query([
                'status' => $content,
            ]),
        ),
    );

    $response = file_get_contents($mastodon_url, false, stream_context_create($options));
    // log error
    if ($response === false) {
        error_log('Error publishing to Mastodon: ' . print_r(error_get_last(), true));
        exit(1);
    }
    return $response;
}

/**
 * Add episode link into file
 */
function mark_as_published($last_episode, $file_path): void
{
    if (($link = $last_episode->link) === null) {
        error_log('Error fetching last episode: ' . print_r(error_get_last(), true));
        exit(1);
    }

    echo "Marking as published: $link\n";

    file_put_contents($file_path, "$link\n", FILE_APPEND);
}

/**
 * Search episode link into file
 */
function is_just_published($last_episode, $file_path): bool
{
    if (($link = $last_episode->link) === null) {
        error_log('Error fetching last episode: ' . print_r(error_get_last(), true));
        exit(1);
    }

	// Create file if it doesn't exist
	if (!file_exists($file_path)) {
		touch($file_path);
	}

    $content = file_get_contents($file_path);

    return str_contains($content, $link);
}

// Main execution
$mastodon_url = 'https://mastodon.uno/api/v1/statuses';
$mastodon_token = getenv('MASTODON_TOKEN');
$config_file = './podcasts.json';

// Load all podcasts configuration
$podcasts = load_podcasts_config($config_file);

echo "Found " . count($podcasts) . " podcast(s) to process\n\n";

// Process each podcast
foreach ($podcasts as $podcast) {
	echo "========================================\n";
	echo "Processing: {$podcast['name']}\n";
	echo "========================================\n";

	$feed_url = $podcast['feed_url'];
	$template = $podcast['template'];
	$file_path = get_tracking_file($podcast['id']);

	// Fetch last episode
    if ($last_episode = fetch_last_episode($feed_url)) {
		echo "Last episode fetched: " . $last_episode->link . "\n";
	} else {
		echo "Error fetching episode for {$podcast['name']}, skipping...\n\n";
		continue;
	}

	// Check if already published
	if (!is_just_published($last_episode, $file_path)) {
		if (publish_to_mastodon($last_episode, $mastodon_url, $mastodon_token, $template)) {
			mark_as_published($last_episode, $file_path);
			echo "✓ Successfully published!\n";
		} else {
			echo "✗ Error publishing to Mastodon\n";
		}
	} else {
		echo "Episode already published\n";
	}

	echo "\n";
}

echo "All podcasts processed!\n";

File: publish.py
import json
import logging
import os
import re
import xml.etree.ElementTree as ET

import requests

from github_state import update_github_variable

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("mastodon")

ITUNES_NS = 'http://www.itunes.com/dtds/podcast-1.0.dtd'


def load_podcasts_config(config_file: str) -> list:
    if not os.path.exists(config_file):
        raise FileNotFoundError(f"File di configurazione non trovato: {config_file}")
    with open(config_file, 'r') as f:
        return json.load(f)


def fetch_last_episode(feed_url: str) -> dict:
    response = requests.get(feed_url)
    response.raise_for_status()

    root = ET.fromstring(response.content)
    item = root.find('./channel/item')

    if item is None:
        raise Exception("Nessun episodio trovato nel feed")

    title = item.findtext('title', '').strip()
    link = item.findtext('link', '').strip()

    if not title or not link:
        raise Exception(f"Titolo o link mancante: {title=} {link=}")

    keywords_el = item.find(f'{{{ITUNES_NS}}}keywords')
    hashtags = ''
    if keywords_el is not None and keywords_el.text:
        hashtags = ' '.join(
            '#' + re.sub(r'\s+', '', kw.strip())
            for kw in keywords_el.text.split(',')
            if kw.strip()
        )

    return {'title': title, 'link': link, 'hashtags': hashtags}


def load_published_urls() -> dict:
    raw = os.environ.get('LAST_PUBLISHED_URLS', '{}')
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        logger.warning("LAST_PUBLISHED_URLS non è JSON valido, uso dict vuoto.")
        return {}


def is_published(link: str, podcast_id: str, published_urls: dict) -> bool:
    return published_urls.get(podcast_id) == link


def publish_to_mastodon(episode: dict, api_url: str, token: str, template: str) -> None:
    content = template \
        .replace('{title}', episode['title']) \
        .replace('{link}', episode['link']) \
        .replace('{hashtags}', episode['hashtags'])

    logger.info(f"Pubblicazione su Mastodon: {content[:80]}...")

    response = requests.post(
        api_url,
        headers={'Authorization': f'Bearer {token}'},
        data={'status': content}
    )

    if response.status_code not in (200, 201):
        raise Exception(f"Errore Mastodon API {response.status_code}: {response.text}")

    logger.info("Post pubblicato con successo!")


if __name__ == "__main__":
    token = os.environ['MASTODON_TOKEN']
    api_url = os.environ.get('MASTODON_API_URL', 'https://mastodon.uno/api/v1/statuses')

    podcasts = load_podcasts_config('./podcasts.json')
    published_urls = load_published_urls()
    logger.info(f"Trovati {len(podcasts)} podcast da processare")

    for podcast in podcasts:
        logger.info(f"=== {podcast['name']} ===")

        episode = fetch_last_episode(podcast['feed_url'])
        logger.info(f"Ultimo episodio: {episode['link']}")

        if is_published(episode['link'], podcast['id'], published_urls):
            logger.info("Episodio già pubblicato, skip.")
            continue

        publish_to_mastodon(episode, api_url, token, podcast['template'])
        published_urls[podcast['id']] = episode['link']
        update_github_variable('LAST_PUBLISHED_URLS', json.dumps(published_urls))

    logger.info("Tutti i podcast processati.")
